specs/UnhandledKeyPressed.md

```diff
@@ -0,0 +1,248 @@
+  /// 3. The event is not recognized or taken over by standard browser shortcuts (like ctrl + F, ctrl + P, etc).
+
+  HRESULT add_UnhandledKeyPressed(
+    [in] ICoreWebView2UnhandledKeyPressedEventHandler* eventHandler,
+    [out] EventRegistrationToken* token);
+
+  /// Removes an event handler previously added with
+  /// `add_UnhandledKeyPressed`.
+
+  HRESULT remove_UnhandledKeyPressed(
+    [in] EventRegistrationToken token);
+}
+
+[uuid(dc83113b-ce7a-47bb-9661-16b03ff8aac1), object, pointer_default(unique)]
+interface ICoreWebView2UnhandledKeyPressedEventHandler : IUnknown {
+
+  /// Provides the event args for the corresponding event.
+
+  HRESULT Invoke(
+      [in] ICoreWebView2Controller* sender,
+      [in] ICoreWebView2UnhandledKeyPressedEventArgs* args);
+}
+
+[uuid(dd30e20c-d1b3-4e51-9bb6-1be7aaa3ce90), object, pointer_default(unique)]
+interface ICoreWebView2UnhandledKeyPressedEventArgs : IUnknown {
+
+  /// The key event type that caused the event to run.
+
+  [propget] HRESULT KeyEventKind([out, retval] COREWEBVIEW2_KEY_EVENT_KIND* keyEventKind);
+
+  /// The Win32 virtual key code of the key that was pressed or released.  It
+  /// is one of the Win32 virtual key constants such as `VK_RETURN` or an
+  /// (uppercase) ASCII value such as `A`.
+  /// use ICoreWebView2UnhandledKeyPressedEventArgs::get_Modifiers()
+  /// instead of Win32 API ::GetKeyState() to verify whether Ctrl or Alt
+  /// is pressed.
+
+  [propget] HRESULT VirtualKey([out, retval] UINT* virtualKey);
+
+  /// The `LPARAM` value that accompanied the window message.  For more
+  /// information, navigate to [WM_KEYDOWN](/windows/win32/inputdev/wm-keydown)
+  /// and [WM_KEYUP](/windows/win32/inputdev/wm-keyup).
+  /// For visual hosting, only scan code and extended key (16-24bit) are provided.
+  /// For other bits, check out the origin message from ::GetMessage().
+
+  [propget] HRESULT KeyEventLParam([out, retval] INT* lParam);
+
+  /// A structure representing the information passed in the `LPARAM` of the
+  /// window message.
+
+  [propget] HRESULT PhysicalKeyStatus(
+      [out, retval] COREWEBVIEW2_PHYSICAL_KEY_STATUS* physicalKeyStatus);
+
+  /// The `Handled` property will not influence the future WebView
+  /// action. It can be used with other UnhandledKeyPressed
+  /// Event handlers.
+
+  [propget] HRESULT Handled([out, retval] BOOL* handled);
+
+  /// Sets the `Handled` property.
+
+  [propput] HRESULT Handled([in] BOOL handled);
+
+  /// Retrieves the status of the keyboard when the key event is triggered.
+  /// Use this instead of ::GetKeyState().
+  /// See COREWEBVIEW2_KEY_PRESSED_FLAG_KIND for details.
+  [propget] HRESULT Modifiers([out, retval] COREWEBVIEW2_KEY_PRESSED_FLAG_KIND* modifiers);
+}
+
+/// Flag bits representing the state of keyboard when a "UnhandledKeyPressed" Event happens.
+[v1_enum]
+typedef enum COREWEBVIEW2_KEY_PRESSED_FLAG_KIND {
```

Comment:
It looks like the values of these match internal Chromium values. Other enum values around keyboard/mouse are based on publicly documented Windows values which are guaranteed not to change in the future. These internal Chromium values do not provider that guarantee, although I don't know how likely it is that they would change. @david-risney, is that an issue?

Reply:
It would be an issue if the chromium value changed and we changed the value in our public API. We cannot change the value in the public API.

For the names, do we already have enums in our WebView2 APIs similar to this? If so, we should be consistent with those names.

If this is a totally new enum and it needs to exist cross-platform then an OS agnostic naming would be OK. If its new and needs to be different on Win and Mac then it would be good to align the names with the existing OS specific names.

Reply:
The values are new and is sync with [upstream] (https://source.chromium.org/chromium/chromium/src/+/main:ui/events/event_constants.h) and is under cross-platform concept. Though it is not expected to change frequently, it is possible for them to change the value (it says it is ok to reorganize the value). For this consideration, maybe we remap the constants from them to WebView @bradp0721 @david-risney? We just simply always map COREWEBVIEW2_KEY_EVENT_FLAG_ALT_DOWN to 1 << 3 regardless of how EF_ALT_DOWN changes its value; We map COREWEBVIEW2_KEY_EVENT_FLAG_ALT_DOWN to EF_ALT_DOWN, not its value.

Reply:
The upstream comment about "It is OK to add values to the middle of this list and/or reorder it" means we have to do an explicit mapping to the COREWEBVIEW2 enum. We cannot just static_cast them and Chromium or Edge could change these values and break our WV2 API.

Instead of having this enum flag, we could also consider adding a GetKeyState function to the event args. The browser side can call GetKeyboardState to get the full keyboard state in a 256 byte array. This would allow the app to query the state of more keys than just the ones we decided to put in the enum. And would make the code the app writes for this event look more similar to the code for AcceleratorKeyPressed. The difference being, one uses ::GetKeyState and one uses event_args->GetKeyState.